GVFS/GVFS.Common/LibGit2RepoPool.cs

```diff
@@ -1,47 +1,66 @@
 ﻿using GVFS.Common.Git;
 using GVFS.Common.Tracing;
 using System;
-using System.Collections.Concurrent;
 using System.Threading;
 
 namespace GVFS.Common
 {
     public class LibGit2RepoPool
     {
-        private const int TryAddTimeoutMilliseconds = 10;
-        private const int TryTakeTimeoutMilliseconds = Timeout.Infinite;
+        private static readonly TimeSpan DefaultRepositoryDisposalPeriod = TimeSpan.FromMinutes(15);
 
-        private readonly BlockingCollection<LibGit2Repo> pool;
+        private readonly Func<LibGit2Repo> createRepo;
         private readonly ITracer tracer;
+        private readonly object sharedRepoLock = new object();
+        private readonly TimeSpan sharedRepositoryDisposalPeriod;
+        private bool stopped;
+        private LibGit2Repo sharedRepo;
+        private Timer sharedRepoDisposalTimer;
+        private int activeCallers;

```

Comment:
Mark `activeCallers` as `volatile` (or use `Interlocked.CompareExchange` in `DisposeSharedRepo`)

Reply:
Actually I see the comparison is inside of `lock (this.sharedRepoLock)`.  I don't know enough about the interaction of `lock` and `Interlocked` to know if the `lock` guarantees that `DisposeSharedRepo` will read the latest value.

Reply:
I can mark it as `volatile` for the purpose of exposing it publicly, but doesn't `Interlocked.Decrement()` do the right thing?

Reply:
We are only reading from `activeCallers` in `DisposeSharedRepo()`, but I can make it `volatile` to make that read safer.

Reply:
> I can mark it as volatile for the purpose of exposing it publicly, but doesn't Interlocked.Decrement() do the right thing?

The `Interlocked.Increment` and `Interlocked.Decrement` will do the right thing when modifying the values, but based on what I've read they alone do not guarantee that reads of the variable from a different thread will see the latest value (see discussion [here](https://stackoverflow.com/questions/3290945/what-to-use-instead-of-interlocked-equals)).

I did just find another discussion [here](https://stackoverflow.com/a/12843092) that mentions the memory barrier created by the `lock` will accomplish the same as `volatile`, and so based on that answer it sounds like the current code is ok as-is (as mentioned above I missed the `lock` on my first pass).  What do you think?

Reply:
Can't hurt to make it `volatile`. The `lock` makes it work, for sure.